tests/test_util.py

```diff
@@ -101,102 +100,66 @@ def get_random_hamiltonian_and_inference(qubits,
+      tf.random.uniform([dim, dim], minval=-val_range, maxval=val_range),
+      tf.complex128)
+  random_imag = tf.cast(
+      tf.random.uniform([dim, dim], minval=-val_range, maxval=val_range),
+      tf.complex128)
+  random_matrix = random_real + random_imag
+  return random_matrix + tf.linalg.adjoint(random_matrix)
 
-  pauli_sum = cirq.PauliSum()
-  for _ in range(num_pauli_terms):
-    pauli_term = random.uniform(coeff_min, coeff_max) * cirq.I(qubits[0])
-    sub_qubits = random.sample(qubits, num_pauli_factors)
-    for q in sub_qubits:
-      pauli_factor = random.choice(paulis)(q)
-      pauli_term *= pauli_factor
-    pauli_sum += pauli_term
-  return pauli_sum
 
+def random_unitary_matrix(num_qubits):
+  """Returns a random unitary matrix.
 
-def generate_pure_random_density_operator(num_qubits):
-  dim = 2**num_qubits
-  unitary = scipy.stats.unitary_group.rvs(dim)
-  u_vec = unitary[:, 0:1]
-  return tf.matmul(u_vec, u_vec, adjoint_b=True)
-
-
-def generate_mixed_random_density_operator(num_qubits, num_mixtures=5):
-  """Generates a random mixed density matrix.
-    Generates `num_mixtures` random quantum states, takes their outer products,
-    and generates a random convex combination of them.
-    NOTE: the states in the mixture are all orthogonal.
-    Args:
-      num_qubits: 2**num_qubits is the size of the density matrix.
-      num_mixtures: the number of pure states in the mixture.  Must be greater
-        than 2**num_qubits.
-    Returns:
-      final_state: The mixed density matrix.
-      prob: The probability of each random state in the mixture.
-    """
-  prob = tf.random.uniform(shape=[num_mixtures])
-  prob = prob / tf.reduce_sum(prob)
-  final_state = tf.zeros((2**num_qubits, 2**num_qubits), dtype=tf.complex128)
-  dim = 2**num_qubits
-  unitary = scipy.stats.unitary_group.rvs(dim)
-  for i in range(num_mixtures):
-    u_vec = unitary[:, i:i + 1]
-    final_state += tf.multiply(
-        tf.cast(prob[i], dtype=tf.complex128),
-        tf.matmul(u_vec, u_vec, adjoint_b=True),
-    )
-  return final_state, prob
+  Uses the property that e^{-iH} is unitary for any Hermitian matrix H.
 
+  Args:
+    num_qubits: Number of qubits on which the matrix acts.
+  """
+  hermitian_matrix = random_hermitian_matrix(num_qubits)
+  return tf.linalg.expm(-1j * hermitian_matrix)
 
-def generate_mixed_random_density_operator_pair(num_qubits, perm_basis=False):
-  num_mixtures = 5
+
+def random_mixed_density_matrix(num_qubits, num_mixtures=5):
+  """Returns a random pure density matrix.
+
+  Applies a common random unitary to `num_mixtures` orthogonal states, then
+  mixes them with random weights.
+
+  Args:
+    num_qubits: Number of qubits on which the matrix acts.
+    num_mixtures: The number of orthogonal pure states to mix.
+
+  Returns:
+    final_state: The mixed density matrix.
+    mixture_probabilities: The probability of each state in the mixture.
+  """
+  pre_probs = tf.random.uniform(shape=[num_mixtures])
```

Comment:
is it fair to say that this can give eigenvalues 0 and so the output matrix can be positive-semidefinite instead of pd? in which case the test should say positive-semidefinite instead of pd?

Reply:
Right, density matrix can have zero eigenvalues (my mental model was too pinned on QHBMs, which ensure positive definiteness).  Your comment also made me realize the minimum on the probabilities should be non-zero, else we could get an all-zeros output matrix.